File: C/tcp-cli-srv/tcp_cli_sum.c
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

#define BUF_SIZE 30
#define SERVER_PORT "9999"

// read a string from stdin protecting buffer overflow
#define GETS(B,S) {fgets(B,S-2,stdin);B[strlen(B)-1]=0;}

int main(int argc, char **argv) {
	int err, sock;
	unsigned long f, i, n, num;
	unsigned char bt;
	char linha[BUF_SIZE];
	struct addrinfo  req, *list;
   
	if(argc!=2) {
        	puts("Server's IPv4/IPv6 address or DNS name is required as argument");
        	exit(1);
        	}

	bzero((char *)&req,sizeof(req));
	// let getaddrinfo set the family depending on the supplied server address
	req.ai_family = AF_UNSPEC;
	req.ai_socktype = SOCK_STREAM;
	err=getaddrinfo(argv[1], SERVER_PORT , &req, &list);
	if(err) {
        	printf("Failed to get server address, error: %s\n",gai_strerror(err)); exit(1); }

	sock=socket(list->ai_family,list->ai_socktype,list->ai_protocol);
	if(sock==-1) {
        	perror("Failed to open socket"); freeaddrinfo(list); exit(1);}

	if(connect(sock,(struct sockaddr *)list->ai_addr, list->ai_addrlen)==-1) {
        	perror("Failed connect"); freeaddrinfo(list); close(sock); exit(1);}
	
	do {
        	do {
			printf("Enter a positive integer to SUM (zero to terminate): ");
			GETS(linha,BUF_SIZE);
			while(sscanf(linha,"%li",&num)!=1 || num<0) {
				puts("Invalid number");
				GETS(linha,BUF_SIZE);
				}
			n=num;
			for(i=0;i<4;i++) {
				bt=n%256; write(sock,&bt,1); n=n/256; }
        		}
        	while(num);
		num=0; f=1; for(i=0;i<4;i++) {read(sock,&bt,1); num=num+bt*f; f=f*256;}
        	printf("SUM RESULT=%lu\n",num);
        	}
	while(num);
	close(sock);
	exit(0);
	}

File: JAVA/tcp-cli-srv/TcpCliSum.java
import java.io.*; 
import java.net.*; 

class TcpCliSum {    
	static InetAddress serverIP;
	static Socket sock;

	public static void main(String args[]) throws Exception {
		if(args.length!=1) {
			System.out.println("Server IPv4/IPv6 address or DNS name is required as argument");
			System.exit(1); }

		try { serverIP = InetAddress.getByName(args[0]); }
		catch(UnknownHostException ex) {
			System.out.println("Invalid server specified: " + args[0]);
			System.exit(1); }

		try { sock = new Socket(serverIP, 9999); }
		catch(IOException ex) {
			System.out.println("Failed to establish TCP connection");
			System.exit(1); }

		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		DataOutputStream sOut = new DataOutputStream(sock.getOutputStream());
		DataInputStream sIn = new DataInputStream(sock.getInputStream());


		String frase;
		long f,i,n,num;
		do {
   			do {
				num=-1;
				while(num<0) {
	  				System.out.print("Enter a positive integer to SUM (zero to terminate): ");
	  				frase = in.readLine();
	 				try { num=Integer.parseInt(frase); }
					catch(NumberFormatException ex) {num=-1;}
	  				if(num<0) System.out.println("Invalid number");
	  				}
				n=num; for(i=0;i<4;i++) {sOut.write((byte)(n%256)); n=n/256; }
				}
   			while(num!=0);
   			num=0; f=1;
			for(i=0;i<4;i++) {num=num+f*sIn.read(); f=f*256;}
   			System.out.println("SUM RESULT = " + num);
   			}
		while(num!=0);
		sock.close();
		} 
	}

